src/cpu/x64/jit_generator.hpp

```diff
@@ -109,19 +111,38 @@ constexpr Xbyak::Operand::Code abi_save_gpr_regs[] = {
 #endif
 };
 
+constexpr Xbyak::Operand::Code abi_param_regs[] = {
 #ifdef _WIN32
-static const Xbyak::Reg64 abi_param1(Xbyak::Operand::RCX),
-        abi_param2(Xbyak::Operand::RDX), abi_param3(Xbyak::Operand::R8),
-        abi_param4(Xbyak::Operand::R9), abi_not_param1(Xbyak::Operand::RDI);
+        Xbyak::Operand::RCX,
+        Xbyak::Operand::RDX,
+        Xbyak::Operand::R8,
+        Xbyak::Operand::R9
 #else
-static const Xbyak::Reg64 abi_param1(Xbyak::Operand::RDI),
-        abi_param2(Xbyak::Operand::RSI), abi_param3(Xbyak::Operand::RDX),
-        abi_param4(Xbyak::Operand::RCX), abi_param5(Xbyak::Operand::R8),
-        abi_param6(Xbyak::Operand::R9), abi_not_param1(Xbyak::Operand::RCX);
+        Xbyak::Operand::RDI,
+        Xbyak::Operand::RSI,
+        Xbyak::Operand::RDX,
+        Xbyak::Operand::RCX,
+        Xbyak::Operand::R8,
+        Xbyak::Operand::R9
 #endif
+};
+
+constexpr Xbyak::Operand::Code abi_not_param_reg =
+#ifdef _WIN32
+        Xbyak::Operand::RDI;
+#else
+        Xbyak::Operand::RCX;
 #endif
 
-} // namespace
+#define abi_param1      Xbyak::Reg64(abi_param_regs[0])
```

Comment:
As far as I understand this solution introduces overhead on creating a `Xbyak::Reg64` object whenever `abi_paramX` is used.
Search for `abi_param1` shows that it's used 263 times. This can affect runtime performance.

Reply:
It shouldn't. First, we're talking about a JIT here and this applies to the generator, not the generated function. The generated content will be exactly the same.

Second, `abi_param1` used to be a global variable before this commit, 32 bits in size, with a specific value that is important only to Xbyak. Now, it's created on demand by the JIT code, with the same value. Since there's no global variable load, this should be even better than before.

Reply:
I understand that this is JIT, I said that because generation time matters and we don't want to slow generation down.

I find it hard to quantify the impact, if benefits are not clear I'd leave it as is to avoid unexpected errors in the future. For example, if a kernel uses a reference to `Xbyak::Reg64` in default member initializer or constructor then it will result in a dangling reference as lifetime of the temporary object created by the proposed macro is not prolonged when it's created within a constructor or as a default initializer.

+ @kwiersch @akharito for comments

Reply:
This is a prvalue, so it can't bind to an lvalue reference anywhere. If the code compiles, then this macro is passed only to parameters by value or by const-ref. And if you're storing a const-ref (64 bits) to a 32-bit variable, you're being wasteful.

Reply:
I think it doesn't compile because of the issue I described: [link](https://dev.azure.com/mkldnn/dnnl/_build/results?buildId=830&view=logs&j=0f2082e4-269f-5e52-9a37-e418fb1c122d&t=6cef451a-a0b5-552a-64fd-0c322932eddc).

Reply:
I'll investigate. It was working with 2.5-rc1 which when I made my changes.

Reply:
Thanks, but I can't fix it because oneDNN currently doesn't build without any of my patches.

Commit e6eef6337
```
/home/tjmaciei/src/oneDNN/src/cpu/x64/jit_uni_x8s8s32x_conv_kernel.hpp: At global scope:
/home/tjmaciei/src/oneDNN/src/common/utils.hpp:64:7: error: expected unqualified-id before ‘const’
   64 |     T(const T &) = delete; \
      |       ^~~~~
/home/tjmaciei/src/oneDNN/src/cpu/x64/jit_uni_x8s8s32x_conv_kernel.hpp:236:5: note: in expansion of macro ‘DNNL_DISALLOW_COPY_AND_ASSIGN’
  236 |     DNNL_DISALLOW_COPY_AND_ASSIGN(jit_uni_x8s8s32x_fwd_kernel<isa>);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tjmaciei/src/oneDNN/src/common/utils.hpp:64:7: error: expected ‘)’ before ‘const’
   64 |     T(const T &) = delete; \
      |      ~^~~~~
/home/tjmaciei/src/oneDNN/src/cpu/x64/jit_uni_x8s8s32x_conv_kernel.hpp:236:5: note: in expansion of macro ‘DNNL_DISALLOW_COPY_AND_ASSIGN’
  236 |     DNNL_DISALLOW_COPY_AND_ASSIGN(jit_uni_x8s8s32x_fwd_kernel<isa>);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Want me to report this as a bug?

Reply:
Never mind, fixed it myself, #1250